crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,322 @@
+
+        let coins = future::join_all(coin_futures)
+            .await
+            .into_iter()
+            .zip(coin_id_page.data)
+            .map(|(r, id)| r.with_internal_context(|| format!("Failed to get object {id}")))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(PageResponse {
+            data: coins,
+            next_cursor: coin_id_page.next_cursor,
+            has_next_page: coin_id_page.has_next_page,
+        })
+    }
+
+    async fn get_coin_id_page(
+        &self,
+        owner: SuiAddress,
+        coin_type_tag: TypeTag,
+        page: Page<BalanceCursor>,
+    ) -> Result<PageResponse<ObjectID, String>, RpcError<Error>> {
+        use coin_balance_buckets::dsl as cb;
+
+        let Self(ctx, _) = self;
+        let mut conn = ctx
+            .reader()
+            .connect()
+            .await
+            .context("Failed to connect to database")?;
+
+        let limit = page.limit;
+
+        let serialized_coin_type =
+            bcs::to_bytes(&coin_type_tag).context("Failed to serialize coin type tag")?;
+
+        // We use two aliases of coin_balance_buckets to make the query more readable.
+        let (candidates, newer) = diesel::alias!(
+            coin_balance_buckets as candidates,
+            coin_balance_buckets as newer
+        );
+
+        // Macros to make the query more readable.
+        macro_rules! candidates {
+            ($($field:ident),*) => {
+                candidates.fields(($(cb::$field),*))
+            }
+        }
+
+        macro_rules! newer {
+            ($($field:ident),*) => {
+                newer.fields(($(cb::$field),*))
+            }
+        }
+
+        // Construct the basic query first to filter by owner, not deleted and newest rows.
+        let mut query = candidates
+            .select((
+                candidates.field(cb::object_id),
+                candidates.field(cb::cp_sequence_number),
+                candidates.field(cb::coin_balance_bucket).assume_not_null(),
+            ))
+            .left_join(
+                newer.on(candidates!(object_id)
+                    .eq(newer!(object_id))
+                    .and(candidates!(cp_sequence_number).lt(newer!(cp_sequence_number)))),
+            )
+            .filter(newer!(object_id).is_null())
+            .filter(candidates!(owner_kind).eq(StoredCoinOwnerKind::Fastpath))
+            .filter(candidates!(owner_id).eq(owner.to_vec()))
+            .filter(candidates!(coin_type).eq(serialized_coin_type))
+            .into_boxed();
+
+        // If the cursor is specified, we filter by it.
+        if let Some(Cursor(BalanceCursor {
+            object_id,
+            cp_sequence_number,
+            coin_balance_bucket,
+        })) = page.cursor
+        {
+            query = query.filter(
+                sql::<Bool>(r#"(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id) < ("#)
```

Comment:
You can drop the raw string if the string contents does not include quotes:
```suggestion
                sql::<Bool>("(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id) < (")
```